src/libraries/System.Linq/tests/LeftJoinTests.cs

```diff
@@ -0,0 +1,450 @@
+                new JoinRec{ name = "Tim", orderID = 95421, total = 9 },
+                new JoinRec{ name = "Robert", orderID = 43421, total = 20 }
+            };
+
+            Assert.Equal(expected, outer.LeftJoin(inner, e => e.custID, e => e.custID, createJoinRec));
+        }
+
+        [Fact]
+        public void OuterSameKeyMoreThanOneElementAndMatches()
+        {
+            CustomerRec[] outer =
+            {
+                new CustomerRec{ name = "Prakash", custID = 98022 },
+                new CustomerRec{ name = "Bob", custID = 99022 },
+                new CustomerRec{ name = "Tim", custID = 99021 },
+                new CustomerRec{ name = "Robert", custID = 99022 }
+            };
+            OrderRec[] inner =
+            {
+                new OrderRec{ orderID = 45321, custID = 98022, total = 50 },
+                new OrderRec{ orderID = 43421, custID = 99022, total = 20 },
+                new OrderRec{ orderID = 95421, custID = 99021, total = 9 }
+            };
+            JoinRec[] expected =
+            {
+                new JoinRec{ name = "Prakash", orderID = 45321, total = 50 },
+                new JoinRec{ name = "Bob", orderID = 43421, total = 20 },
+                new JoinRec{ name = "Tim", orderID = 95421, total = 9 },
+                new JoinRec{ name = "Robert", orderID = 43421, total = 20 }
+            };
+
+            Assert.Equal(expected, outer.LeftJoin(inner, e => e.custID, e => e.custID, createJoinRec));
+        }
+
+        [Fact]
+        public void NoMatches()
+        {
+            CustomerRec[] outer =
+            {
+                new CustomerRec{ name = "Prakash", custID = 98022 },
+                new CustomerRec{ name = "Bob", custID = 99022 },
+                new CustomerRec{ name = "Tim", custID = 99021 },
+                new CustomerRec{ name = "Robert", custID = 99022 }
+            };
+            OrderRec[] inner =
+            {
+                new OrderRec{ orderID = 45321, custID = 18022, total = 50 },
+                new OrderRec{ orderID = 43421, custID = 29022, total = 20 },
+                new OrderRec{ orderID = 95421, custID = 39021, total = 9 }
+            };
+            JoinRec[] expected =
+            {
+                new JoinRec{ name = "Prakash", orderID = 0, total = 0 },
+                new JoinRec{ name = "Bob", orderID = 0, total = 0 },
+                new JoinRec{ name = "Tim", orderID = 0, total = 0 },
+                new JoinRec{ name = "Robert", orderID = 0, total = 0 }
+            };
+
+            Assert.Equal(expected, outer.LeftJoin(inner, e => e.custID, e => e.custID, createJoinRec));
+        }
+
+        [Fact]
+        public void ForcedToEnumeratorDoesntEnumerate()
+        {
+            var iterator = NumberRangeGuaranteedNotCollectionType(0, 3).LeftJoin(Enumerable.Empty<int>(), i => i, i => i, (o, i) => i);
+            // Don't insist on this behaviour, but check it's correct if it happens
```

Comment:
I'm not sure I fully understand what this test is doing. If it's checking that `LeftJoin` doesn't eagerly enumerate either of the inputs, the assertion doesn't necessarily validate that. Would it make sense to pass throwing enumerators (i.e. have the calls to `GetEnumerator()` throw) and ensure that `LeftJoin` still returns a result?

Reply:
Yeah, I was wondering about that as well - but note that this test is duplicated as-is across many/all LINQ operators. this is a copy-paste from JoinTests, where the same problem exists, I think...

I'm happy to help improve this, but maybe as a separate PR to keep this one focused?

Reply:
I think it's fine to keep as-is